### ISSUE TITLE
Keyboard does not detect prop change of function props

### ISSUE THREAD

**Issue by jmadelaine** [hodgef/react-simple-keyboard#545]
**React-simple-keyboard version**
2.2.83

## Describe the bug

Once the Keyboard component is initialized, the `onKeyPress` prop can not be changed and the Keyboard continues to use stale props (the original `onKeyPress` function used to initialize the Keyboard).

## Example

CodeSandbox: https://codesandbox.io/s/modern-smoke-5jo8d

The example works in the following way:
- there are 2 strings
- when clicking a button or pressing a virtual keyboard key, we write to whichever string is the shortest. Therefore, the string that gets written to will switch with every render.

The button in the example works correctly, and allows the `onClick` function to be updated between renders, therefore writing to the shortest string each time.

The Keyboard in the example does not work correctly, and the value of `onKeyPress` does not change for the life of the component, therefore always writing to the initial string.

## Source of bug

The bug is in the function that is responsible for detecting a prop change:

https://github.com/hodgef/react-simple-keyboard/blob/master/src/lib/services/Utilities.js

```js
export const propsChanged = (prevProps, props) => {
  const cleanProps = sourceObj =>
    JSON.stringify({
      ...sourceObj,
      stateToIgnore: null
    });

  return cleanProps(props) !== cleanProps(prevProps);
};
```

When the props are 'cleaned', any function props are removed, therefore changes to function props will never be detected.

Running the following code:

```js
const props = {
  someNumber: 5,
  onKeyPress: () => {
    console.log("You pressed a key");
  }
};

const cleanProps = sourceObj =>
  JSON.stringify({
    ...sourceObj,
    stateToIgnore: null
  });

console.log("before clean:", props);
console.log("after clean:", cleanProps(props));
```

gives the output:

```
before clean: { someNumber: 5, onKeyPress: [Function: onKeyPress] }
after clean: {"someNumber":5,"stateToIgnore":null}
```

Is there a reason that we are using this cleaning function to compare props, or indeed, comparing props at all?

React provides `memo` which is a way to prevent re-render on specific prop changes.

**Comment by hodgef** [comment]
Hey @jmadelaine,

Thanks a lot for flagging this issue in such detail. This has been reported before in https://github.com/hodgef/react-simple-keyboard/issues/423 which I had mistakenly dismissed.

Here's some context:

This library is a wrapper of my [vanilla-js library](https://github.com/hodgef/simple-keyboard). When props change I call the [`setOptions` function](https://github.com/hodgef/simple-keyboard/blob/master/src/lib/components/Keyboard.js#L1203) from the core library. This triggers a re-processing of the keyboard buttons. So the props check is to ensure the props have actually changed.

Without this check, `setOptions` keeps getting called. This means the keyboard buttons are always re-rendered. In the core-js library this doesn't happen, but in React every button click triggers a `setState`, which means the view is always re-rendered. I hope that explains it.

I have to admit I have never used `React.memo` before. I will look into it today and hopefully release a fix tonight.

Thanks again for taking the time to explain this issue. I've been debugging this earlier due to an user raising it and I couldn't find the cause. I believe you answered the [user's question](https://stackoverflow.com/questions/61334337/react-usestate-broken-state-is-not-updated-correctly) on SO, and basically uncovered this. I can't thank you enough!

I'll keep updating this thread as this progresses.

Best,
Francisco Hodge

**Comment by jmadelaine** [comment]
Hey @hodgef , thanks for a quick response and explanation!

**Comment by hodgef** [comment]
This is now fixed as of version `2.3.0`. [The fix](https://github.com/hodgef/react-simple-keyboard/commit/79e6ad7c5f540706bf206e862df1e739e445584e) is basically a shallow compare, which is not super robust but address this issue.

I'll keep monitoring this ticket should any side-effects arise.

Thanks again for your help!
Francisco Hodge
